### PR TITLE
Sofia edits

### PR DESCRIPTION
**Pull request by SofiaBariami**
Implementation of geometric combining rules for SOMD

### CODE REVIEW

**Comment by chryswoods** [comment]
Thanks for submitting a good pull request. Your code looks great.

Just a quick question. Why do you use CombRules as the name? The rest of the code uses CombiningRules, with functions called setCombiningRules, e.g. https://github.com/michellab/Sire/blob/devel/corelib/src/libs/SireMM/cljfunction.h

The property name is “combiningRules”, e.g. see https://github.com/michellab/Sire/blob/devel/corelib/src/libs/SireMM/cljfunction.cpp

(function CLJFunctionPtr CLJFunction::setProperty(const QString &name, const Property &value) const)

If possible it would be great to use the same name and function naming scheme as this will make it less confusing for users.

Also, I believe you have mostly followed it, but as a reminder, here is the coding style document.
https://github.com/michellab/Sire/blob/devel/CODING_STYLE.md

File: wrapper/Tools/OpenMMMD.py [hunk 1]
```diff
@@ -1077,7 +1077,7 @@ def setupForceFieldsFreeEnergy(system, space):
     else:
         system.setProperty("switchingFunction", NoCutoff())
 
-    system.setProperty("CombRules", VariantProperty(combining_rules.val))
+    system.setProperty("CombiningRules", VariantProperty(combining_rules.val))
```
Comment (chryswoods) [inline wrapper/Tools/OpenMMMD.py]: should be "combiningRules" - sorry, small "c"

File: wrapper/Tools/OpenMMMD.py [hunk 2]
```diff
@@ -408,7 +408,7 @@ def setupForcefields(system, space):
 
     system.setProperty("space", space)
     system.setProperty("switchingFunction", CHARMMSwitchingFunction(cutoff_dist.val))
-    system.setProperty("CombRules", VariantProperty(combining_rules.val))
+    system.setProperty("CombiningRules", VariantProperty(combining_rules.val))
```
Comment (chryswoods) [inline wrapper/Tools/OpenMMMD.py]: Sorry - "combiningRules" with a small "c". Sorry to be picky - need to match property name in other parts of the code

Comment (SofiaBariami) [reply wrapper/Tools/OpenMMMD.py]: Sorry about that! My mistake. I'm fixing it now

Comment (SofiaBariami) [reply wrapper/Tools/OpenMMMD.py]: hi @chryswoods ! can you see the commits that I did or should I "merge the pull request" ?

**Comment by chryswoods** [comment]
Thanks for making the changes. The CI is showing errors compiling for OS X (click on "Details" above for the OS X build and you can then see that there is a build error (see picture or error copied below)

```
(/Users/vsts/agent/2.146.0/work/1/s/corelib/src/libs/SireMove/openmmfrenergyst.cpp:133:9: error: no member named 'Rules' in 'SireMove::OpenMMFrEnergyST')
```

screenshot 2019-02-15 at 16 28 37

Please fix the compile error. Sorry that this seems like an annoying process - it is the best way to manage lots of people separately editing code. Once this passes for Linux and OS X I will accept the request.

File: wrapper/Tools/OpenMMMD.py (hunk 2 above)
Comment (chryswoods) [reply wrapper/Tools/OpenMMMD.py]: Yes, I can see the commits. Thanks for checking :-). No, policy is you can't merge your own pull request. As I've commented below, the change has introduced a compile error for OS X. Once fixed I will review again and then merge.

**Comment by chryswoods** [comment]
Thanks - I believe this will work so will merge the pull request as soon as the tests are complete (or after teaching if the CI/CD takes too long ;-))

**Comment by chryswoods** [comment]
Thanks - you are now free to delete the branch if you want

**Comment by SofiaBariami** [comment]
Thank you very much for your help Christopher! I deleted the branch. 
You will probably see more code of mine from now on (that sounded like a threat)!  :)  
Have a nice rest of the day!

**Comment by jmichel80** [comment]
The new code in openmmfrenergyst.cpp duplicates a large amount of the previous code in order to introduce a minor change to the nonbondedforce expressions e.g.


    if (flag_combRules == ARITHMETIC)
    {
(ca. 500 lines of code)
    else // IF GEOMETRIC
(ca. 500 very similar lines of code, different only in expressions used for  'sigma_avg' )
   }

Instead of doing that you should call the constructors of the CustomForces using a std::string constructed according to the combining rules flags. For instance

```
    if (flag_cutoff == NOCUTOFF)
    {

        if (coulomb_power > 0)
        { //This is necessary to avoid nan errors on the GPUs platform caused by the calculation of 0^0

            std::string energybase = ""(1.0 - isSolvent1 * isSolvent2 * SPOnOff) * (Hcs + Hls);"
             "Hcs = (lambda^n) * 138.935456 * q_prod/sqrt(diff_cl+r^2);"
              "diff_cl = (1.0-lambda) * 0.01;"
               "Hls = 4.0 * eps_avg * (LJ*LJ-LJ);"
              "LJ=((sigma_avg * sigma_avg)/soft)^3;"
             "soft=(diff_lj*delta*sigma_avg + r*r);"
               "diff_lj=(1.0-lambda) * 0.1;"
               "lambda = Logic_lam * lam + Logic_om_lam * (1.0-lam) + Logic_mix_lam * max(lam,1.0-lam) + Logic_hard;"
               "Logic_hard = isHD1 * isHD2 * (1.0-isTD1) * (1.0-isTD2) * (1.0-isFD1) * (1.0-isFD2);"
              "Logic_om_lam = max((1.0-isHD1)*(1.0-isHD2)*isTD1*isTD2*(1.0-isFD1)*(1.0-isFD2), B_om_lam);"
                "B_om_lam = max(isHD1*(1.0-isHD2)*isTD1*(1.0-isTD2)*(1.0-isFD1)*(1.0-isFD2), C_om_lam);"
                   "C_om_lam = max(isHD1*(1.0-isHD2)*(1.0-isTD1)*isTD2*(1.0-isFD1)*(1.0-isFD2) , D_om_lam);"
                  "D_om_lam = max((1.0-isHD1)*isHD2*isTD1*(1.0-isTD2)*(1.0-isFD1)*(1.0-isFD2), E_om_lam);"
              "E_om_lam = (1.0-isHD1)*isHD2*(1.0-isTD1)*isTD2*(1.0-isFD1)*(1.0-isFD2);"
             "Logic_lam = max((1.0-isHD1)*(1.0-isHD2)*(1.0-isTD1)*(1.0-isTD2)*isFD1*isFD2, B_lam);"
            "B_lam = max(isHD1*(1.0-isHD2)*(1.0-isTD1)*(1.0-isTD2)*isFD1*(1.0-isFD2), C_lam);"
            "C_lam = max(isHD1*(1.0-isHD2)*(1.0-isTD1)*(1.0-isTD2)*(1.0-isFD1)*isFD2 , D_lam);"
               "D_lam = max((1.0-isHD1)*isHD2*(1.0-isTD1)*(1.0-isTD2)*isFD1*(1.0-isFD2), E_lam);"
        "E_lam = (1.0-isHD1)*isHD2*(1.0-isTD1)*(1.0-isTD2)*(1.0-isFD1)*isFD2;"
              "Logic_mix_lam = max((1.0-isHD1)*(1.0-isHD2)*isTD1*(1.0-isTD2)*isFD1*(1.0-isFD2), B_mix);"
               "B_mix = max((1.0-isHD1)*(1.0-isHD2)*isTD1*(1.0-isTD2)*(1.0-isFD1)*isFD2, C_mix);"
        "C_mix = max((1.0-isHD1)*(1.0-isHD2)*(1.0-isTD1)*isTD2*isFD1*(1.0-isFD2) , D_mix);"
         "D_mix= (1.0-isHD1)*(1.0-isHD2)*(1.0-isTD1)*isTD2*(1.0-isFD1)*isFD2;"
              "q_prod = (qend1 * lam+(1.0-lam) * qstart1) * (qend2 * lam+(1.0-lam) * qstart2);"
             "eps_avg = sqrt((epend1*lam+(1.0-lam)*epstart1)*(epend2*lam+(1.0-lam)*epstart2));"" 

                 if (flag_combRules == ARITHMETIC)
                      {
                            energybase.append("sigma_avg = 0.5*((sigmaend1*lam+(1.0-lam)*sigmastart1)+(sigmaend2*lam+(1.0-lam)*sigmastart2))")
                       }
                   else if (flag_combRules == GEOMETRIC)
                      {
                            energybase.append("sigma_avg = sqrt((sigmaend1*lam+(1.0-lam)*sigmastart1)*(sigmaend2*lam+(1.0-lam)*sigmastart2))")
                      }
                   custom_force_field = new OpenMM::CustomNonbondedForce(energybase) ;
(..)
```


Also for geometric combining rules the the sigma terms should be combined as the eps terms for all types of interactions

So custom_intra_14_todummy with geometric combining rules should look like eps_avg, i.e.

```

  "eps_avg = sqrt((1-lamtd)*(1-lamtd)*eaend + lamtd*lamtd*eastart + lamtd*(1-lamtd)*emix);"
   "sigma_avg = sqrt((1-lamtd)*(1-lamtd)*saend + lamtd*lamtd*sastart + lamtd*(1-lamtd)*samix);

```
NOT

`sigma_avg = (1-lamtd)*saend + lamtd*sastart;
`
Note that this requires to add a new PerBondParameter to the various custom_intra_14_clj forces to support the parameter ''samix'' in analogy with ''emix'' see around l1720 and line 3053-3074, otherwise the energies should not agree at lambda values different from 0 and 1 between Sire and SOMD.

**Pull request merged**